File: psbt-support/psbt-builder.mjs
/**
 * PSBT Builder (Node.js)
 *
 * Fetches a bump plan from FastPath and constructs an unsigned PSBT
 * ready for wallet signing (Sparrow, Electrum, Bitcoin Core).
 *
 * Usage:
 *   EVM_PRIVATE_KEY=0x... node psbt-support/psbt-builder.mjs <txid> [targetSatVb]
 */

import { writeFileSync } from "fs";
import * as bitcoin from "bitcoinjs-lib";
import * as ecc from "tiny-secp256k1";
import { createFastPathX402Client } from "../x402-optimizer-middleware.js";

bitcoin.initEccLib(ecc);

const txid = process.argv[2];
const targetSatVb = Number(process.argv[3] || 25);

if (!txid) {
  console.error("Usage: node psbt-support/psbt-builder.mjs <txid> [targetSatVb]");
  process.exit(1);
}

if (!process.env.EVM_PRIVATE_KEY) {
  console.error("Set EVM_PRIVATE_KEY to a bot wallet funded with USDC.");
  process.exit(1);
}

async function buildPsbt() {
  console.log(`Fetching bump plan for ${txid} (Target: ${targetSatVb} sat/vB)...`);

  const client = createFastPathX402Client({
    apiBase: process.env.FASTPATH_API_BASE || "https://api.nativebtc.org",
    privateKey: process.env.EVM_PRIVATE_KEY,
  });

  const res = await client.bumpPlan(txid, targetSatVb);
  const data = res.data;

  if (!data?.bump?.available) {
    console.log(`\nNo bump plan available for ${txid}.`);
    console.log(`Reason: ${data?.reason || "Transaction does not require or support RBF/CPFP."}`);
    return;
  }

  const inputDetails = data.inputDetails || data.bump?.inputDetails;

  if (!inputDetails || inputDetails.length === 0) {
    console.error("\nError: Node response missing inputDetails required to build PSBT.");
    console.error("Node must have txindex=1 enabled to resolve parent UTXOs.");
    process.exit(1);
  }

  console.log("\nConstructing unsigned PSBT...");

  const psbt = new bitcoin.Psbt({ network: bitcoin.networks.bitcoin });

  // Add inputs from resolved details
  for (const input of inputDetails) {
    const inputData = {
      hash: input.txid,
      index: input.vout,
    };

    if (input.nonWitnessUtxo) {
      inputData.nonWitnessUtxo = Buffer.from(input.nonWitnessUtxo, "hex");
    } else if (input.witnessUtxo) {
      inputData.witnessUtxo = {
        script: Buffer.from(input.witnessUtxo.scriptHex, "hex"),
        value: BigInt(input.witnessUtxo.value),
      };
    }

    psbt.addInput(inputData);
  }

  // Add outputs specified in the bump plan
  const outputs = data.bump?.outputs || [];
  for (const output of outputs) {
    psbt.addOutput({
      address: output.address,
      value: BigInt(output.valueSats),
    });
  }

  const psbtBase64 = psbt.toBase64();
  const filename = "bump_replacement.psbt";

  writeFileSync(filename, psbtBase64);

  console.log("\n✅ Unsigned PSBT created successfully!");
  console.log(`Saved to: ${filename}`);
  console.log("\nImport this file into Sparrow, Electrum, or Bitcoin Core to sign and broadcast.");
}

buildPsbt().catch((err) => {
  console.error("\nPSBT construction failed:", err.message || err);
  process.exit(1);
});

File: psbt-support/psbt_builder.py
"""
PSBT Builder (Python)

Fetches a bump plan from FastPath and constructs an unsigned PSBT file.

Usage:
  EVM_PRIVATE_KEY=0x... python psbt-support/psbt_builder.py <txid> [target_sat_vb]
"""

import os
import sys
import json
import base64
import requests

API_BASE = os.getenv("FASTPATH_API_BASE", "https://api.nativebtc.org")

if len(sys.argv) < 2:
    print("Usage: python psbt-support/psbt_builder.py <txid> [target_sat_vb]")
    sys.exit(1)

txid = sys.argv[1]
target_sat_vb = int(sys.argv[2]) if len(sys.argv) > 2 else 25
private_key = os.getenv("EVM_PRIVATE_KEY")

if not private_key:
    print("Set EVM_PRIVATE_KEY to a bot wallet funded with USDC.")
    sys.exit(1)

def get_bump_plan(txid, target_sat_vb):
    url = f"{API_BASE}/v1/template/bump-plan"
    payload = {"txid": txid, "targetSatVb": target_sat_vb}
    headers = {"content-type": "application/json"}
    
    # Probe or execute request
    res = requests.post(url, json=payload, headers=headers)
    
    if res.status_code == 402:
        print("Payment required. Use x402 client wrapper or node environment for automated payment.")
        sys.exit(1)
        
    if not res.ok:
        print(f"API Error {res.status_code}: {res.text}")
        sys.exit(1)
        
    return res.json()

def main():
    print(f"Fetching bump plan for {txid}...")
    data = get_bump_plan(txid, target_sat_vb)
    
    bump = data.get("bump", {})
    if not bump.get("available"):
        print(f"No bump plan available for {txid}.")
        return

    psbt_b64 = bump.get("psbtBase64") or data.get("psbtBase64")
    
    if not psbt_b64:
        print("Error: Bump plan returned no raw PSBT data.")
        sys.exit(1)
        
    filename = "bump_replacement.psbt"
    with open(filename, "w") as f:
        f.write(psbt_b64)
        
    print(f"\n✅ Unsigned PSBT saved to {filename}")

if __name__ == "__main__":
    main()
